### ISSUE TITLE
Weak event.add

### ISSUE THREAD

**Issue by sebix** [certtools/intelmq#778]
In practice we often need to add values to events if they are valid. So something like this would be good:

    event.add('source.ip', '127.0.0.', ignore=True)

In this case, the data is not added. We can additionally return False, so we get direct feedback in the bot if the add was successful.

Better names for the parameter are welcome :)

**Comment by aaronkaplan** [comment]
> On 16 Nov 2016, at 15:45, Sebastian <notifications@github.com> wrote:
> 
> In practice we often need to add values to events if they are valid. So something like this would be good:
> 
> event.add('source.ip', '127.0.0.', ignore=True)
> 
> In this case, the data is not added. We can additionally return False, so we get direct feedback in the bot if the add was successful.

You mean if the add would have been successful?

So it's basically a check (syntax, semantic) if an add would work?


> 
> Better names for the parameter are welcome :)
> 

event.test_value(...) ?

**Comment by sebix** [comment]
Currently the code uses e.g. URL.is_valid(value) with URL from the
harmonization library. However, this does not perform all checks for
this field e.g. regex. And this can be shorter.
So instead of:
if FQDN.is_valid(row[3]):
    event.add('source.fqdn', row[3])
we could have
event.add('source.fqdn', row[3], weak=True)

Or in case a field has either FQDN or IP
if not event.add('source.ip', row[3], weak=True)
    event.add('source.fqdn', row[3])

**Comment by dmth** [comment]
what about `event.add('barbaz', dryrun=True)` or `event.add('barbaz', testonly=True)`?

**Comment by aaronkaplan** [comment]
> On 16 Nov 2016, at 16:48, Dustin Demuth <notifications@github.com> wrote:
> 
> what about event.add('barbaz', dryrun=True) or event.add('barbaz', testonly=True)?
> 

sounds good in case you want to achieve this behaviour in general
-> but please also let's document this.

**Comment by aaronkaplan** [comment]
> On 16 Nov 2016, at 16:00, Sebastian <notifications@github.com> wrote:
> 
> Currently the code uses e.g. URL.is_valid(value) with URL from the
> harmonization library. However, this does not perform all checks for
> this field e.g. regex. And this can be shorter.
ok

> So instead of:
> if FQDN.is_valid(row[3]):
> event.add('source.fqdn', row[3])
> we could have
> event.add('source.fqdn', row[3], weak=True)
> 

or (as written before): dryrun=True


In particular, I want to point out the faup library for URL checking.

https://github.com/stricaud/faup

**Comment by sebix** [comment]
dryrun is good to do only the check. For this we could also do `event.check` or `event.is_valid`.

What I wanted to do originally is a function which adds the value if it's valid and *not* raise an exception if it's invalid, because the field is not important (which is default currently). So we can do the if-valid-then-add operation with one command.

**Comment by dmth** [comment]
> What I wanted to do originally is a function which adds the value if it's valid and not raise an exception if it's invalid, because the field is not important (which is default currently). So we can do the if-valid-then-add operation with one command.

Aha! What about `event.add('barbaz', [silent|quiet]=True) `

**Comment by aaronkaplan** [comment]
> On 16 Nov 2016, at 17:32, Dustin Demuth <notifications@github.com> wrote:
> 
> What I wanted to do originally is a function which adds the value if it's valid and not raise an exception if it's invalid, because the field is not important (which is default currently). So we can do the if-valid-then-add operation with one command.
> 
> Aha! What about event.add('barbaz', [silent|quiet]=True)

silent suggests to the reader that it is being done but no logging or similar will happen.

If I understood sebix correctly, what he is suggesting to have a "dry-run" feature (try adding the data to the event but don't really write it there)

IMHO it is very very important how we name things.
Because it sets expectations (implicit expectations) on the readers' / programmers side.
If you remember some of the discussions we had about the DHO, ... they all go back to unclear meanings and wording.

**Comment by sebix** [comment]
On 11/16/2016 06:08 PM, AaronK wrote:
> silent suggests to the reader that it is being done but no logging or
> similar will happen.
But silent would mean "no exception if invalid".
> If I understood sebix correctly, what he is suggesting to have a
> "dry-run" feature (try adding the data to the event but don't really
> write it there)
Both, the dry-run and siltent are useful I think. Originally I thought
only about the silent mode, but dry run is also necessary.

dry_run is a very clear and intuitive parameter, I think agree on this
one. But 'silent' is not. A clear one would be ignore_if_invalid=True.
However a bit log. But better long than vague.

**Comment by cgi1** [comment]
`dry_run` +1

**Comment by sebix** [comment]
I don't understand why `dry_run` is easy to understand. `event.add(..., dry_run=True)` makes clear the action is *not performed*, but it does say nothing about what it *does* and what it is intended *to be used for*. While `is_valid` implies that nothing is added *and* a check is performed.

E.g.:

    >>> event.is_valid('source.ip', '127.0.0.x')
    False
    >>> event.is_valid('source.ip', '127.0.0.1')
    True

For the second usecase (https://github.com/certtools/intelmq/issues/778#issuecomment-260967498) I still don't know if the parameters `weak=True` and `ignore_if_invalid=True` are a good idea/choice.

E.g.

    >>> event.add('source.ip', '127.0.0.x', weak=True)
    False
    # no error, not added
    >>> event.add('source.ip', '127.0.0.1', weak=True)
    True
    # added

**Comment by robcza** [comment]
For the second use case I would suggest ```safe```, ```cautious``` or ```opted```
I consider this feature very handy, thanks for this idea @sebix

**Comment by SYNchroACK** [comment]
#### From my point of view there are 3 issues on this thread:

1. Mechanism to fully and easily control of exceptions being raised due an invalid value which is trying to be added to an event-key
2. Name of the parameter to enable or disable this mechanism
3. Solve the regex parameter from harmonization that needs to be applied.

#### My perspectives for each one:
1. fully agree, excellent idea! @sebix, I would say that is even better than the scenario that you present. With this mecanism you can do the following:

```
report_example = {
    "domain": "www.intelmq.org",
    "ip_address": "127.0.",
    "url": "http://www.intelmq.org/example.php" 
}

for key, value in report_example.iteritems():

    if key == "URL":
        event.add('source.url', value)

    if key == "domain" OR key == "ip_address":
        event.add('source.fqdn', value, raise_failure=False)
        event.add('source.ip', value, raise_failure=False)
```

2. My suggestion is `raise_failure`. The dafault value will be always True (the current behavior) and in some scenarios like @sebix already present the developer can specify False in order to not have the bot raising a failure which will trigger the error handling procedure configured.

3. I see harmonization.py as an independent package which should be capable to load the harmonization.conf file and perform autonomously all checks needed, including the regex which is part of the validation process and SHOULD NOT added to the message.py library. I'm not saying that is an easy task but conceptually IMHO this is the path.
.
**TL;DR** another issue and your idea should not be a way to solve that issue. (sorry if I miss something from your perspective)

**Comment by wagner-certat** [comment]
My proposal is in #841

> I see harmonization.py as an independent package which should be capable to load the harmonization.conf file and perform autonomously all checks needed, including the regex which is part of the validation process and SHOULD NOT added to the message.py library. I'm not saying that is an easy task but conceptually IMHO this is the right way.

Could be part of #671 and #616 (milestone 1.1)

**Comment by SYNchroACK** [comment]
> My proposal is in #841

Perfect! This should be part of 1.0

> I see harmonization.py as an independent package which should be capable to load the harmonization.conf file and perform autonomously all checks needed, including the regex which is part of the validation process and SHOULD NOT added to the message.py library. I'm not saying that is an easy task but conceptually IMHO this is the right way.
> Could be part of #671 and #616 (milestone 1.1)

Yes, agree. Should be part of 1.1.

Thank you @sebix
